### ISSUE TITLE
BUG: v3.9.20 laggy, and using lots of CPU

### ISSUE THREAD

**Issue by pcolby** [beekeeper-studio/beekeeper-studio#1703]
**Describe the bug**

Beekeeper Studio CE 3.9.20 is using 60 ~ 90% CPU continuously, and is very laggy.

**To Reproduce**

1. Connect to a (PostgreSQL) database.
2. Try typing / editing any SQL.

**Expected behavior**

Usually its nice and fast.

**Version Information (please complete the following information):**
 - OS:  Ubuntu 22.04.3 LTS
 - App Version 3.9.20 (Community Edition)
 - PostgreSQL 13.8

**Additional context**

Would love to help diagnose the issue. Let me know if there's anything I can try, or more info I can provide.

Thanks!

**Comment by pcolby** [comment]
In case it helps, running as:

```
beekeeper --enable-logging --log-file=log.txt --log-level=10 -v=2
```

gives the following console output:

```
[1:0920/161750.999574:INFO:cpu_info.cc(53)] Available number of cores: 8
[772089:0920/161750.999569:INFO:cpu_info.cc(53)] Available number of cores: 8
[1:0920/161750.999710:VERBOSE1:zygote_main_linux.cc(217)] ZygoteMain: initializing 0 fork delegates
[772089:0920/161750.999657:VERBOSE1:zygote_main_linux.cc(217)] ZygoteMain: initializing 0 fork delegates
16:17:54.310                     › initializing background
running migrations
[772117:0920/161755.309020:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
```

And wrote the following `log.txt` file: [log.txt](https://github.com/beekeeper-studio/beekeeper-studio/files/12669219/log.txt)

**Comment by azmy60** [comment]
Hi @pcolby , thanks for reporting this! 

Just need a little clarification, does it start happening when you're using the autocomplete? Specifically after typing dot `.` or `from ` followed by space?

I assume this is due to the new schema autocomplete in this PR #1610.

**Comment by pcolby** [comment]
Hi @azmy60,

The lagginess is obvious when typing, or moving the cursor around the SQL editor. For example, if I blank the editor entirely, then just hold down one letter (eg `s`) for a few seconds, it will take another several seconds before they appear in the editor.

Also, I notice that the CPU jumps, and stays very high (40 ~ 80%) as soon as I connect to the database, even if the editor was completely empty and is left empty.

The CPU will stay in that elevated range as long as the window is visible. For example, the CPU will remain high:
* if the current selection is moved out of the editor (eg clicking elsewhere in the app); or
* if the focus is moved to a completely different application, but Beekeeper is still visible.

But Beekeeper's CPU will drop to 0% if:
* I minimise the window; or
* I place another window over the top of Beekeeper, such that Beekeeper is entirely covered. Even just small part of the status bar being visible is enough to keep the CPU very high.

Hope some of that helps. Happy to try other tests too :)

Thanks!

**Comment by pcolby** [comment]
Just doing some more experiments, I moved my `~/.local/beekeeper-studio` folder temporarily, and started Beekeeper afresh. It still has the same issue.  However, I see now that the CPU does drop if the mouse stops moving over the query editor.  In case that helps at all.

**Comment by davidkaufman** [comment]
Also does it happen when you connect to other databases, or just this one?  And if not, do you have a huge number of tables, schemas or columns in the db?

**Comment by pcolby** [comment]
So I did a but more testing, and found something interesting. But first I'll quickly answer all the questions.

> how did you install Beekeeper? Is it the Snap version?

Just downloaded `Beekeeper-Studio-3.9.20.AppImage` and `chmod u+x` it. Not the Snap version.

> Are you running Wayland or X?

x11

> Are you using the Gnome desktop, or something else?

I'm using lxqt.

> Are any other electron apps having this problem? (eg VSCode, GitKraken, Slack)

Nope, VSCode and Slack are both fine.

> does it happen when you connect to other databases, or just this one?

This happens with all the databases I use, but they're all essentially clones of each other, so more or less the same.

I tried creating an empty SQLite database as a quick test, and it worked fine.

I then did some more playing, and accidentally discovered that the performance is great if I just close the `public` schema in the tree view.

The database/s in question have 7 schemas. Beekeeper is reporting "1770 entities", but a huge number of those are PostGIS functions, which (in my case) are all in the `public` schema.  If I expand the other 6 schemas, and leave `public` closed, then query editor the performance is great - even though the query editor's auto-complete still suggests things from the `public` schema.  But as long as I have the `public` schema open, then the query editor performance becomes unmanageable.

Not sure why the query editor would be affected this way by the treeview, especially since the view doesn't seem to affect auto-complete suggestions.  But I can definitely understand a huge number of entities affecting performance... it might be appropriate to filter out PostGIS functions from the treeview? or not? 🤷🏻‍♂️

Anyway, I'm extremely glad to at least have a workaround for now: simply closing the `public` schema in the tree 😄

**Comment by rathboma** [comment]
Very interesting. @azmy60 can you take a look at this? Might be related to auto complete?

**Comment by azmy60** [comment]
No problem!

**Comment by azmy60** [comment]
I'm able to reproduce this with 2000 tables in public schema, but it also happens in other tabs too, even when there is no query tab, e.g. editing a column in Table Structure, typing a filter value in Table View. It goes back to normal when I close the public schema tree view as you mentioned @pcolby.

@pcolby In your device, does it happen with other tabs too? not just the query tab?

**Comment by davidkaufman** [comment]
@azmy60 you might want to check if it's not just the number of entries in the schema that's causing the performance hit, but maybe some other resource is constrained like db activity.  If the db is constantly being queried, that might be slowing it down, and the function could just cache the results in memory?

**Comment by azmy60** [comment]
Oh you're right! Filter builder will query some data every time you type. So both the treeview and the constant querying need some fixes.

**Comment by azmy60** [comment]
Closing this in favor of #1771. This fix was released in 4.1.x.
